// x402-hono/dist/esm/index.mjs
// src/index.ts
import { getAddress } from "viem";
import { exact } from "x402/schemes";
import {
  computeRoutePatterns,
  findMatchingPaymentRequirements,
  findMatchingRoute,
  getPaywallHtml,
  processPriceToAtomicAmount,
  toJsonSafe
} from "x402/shared";
import {
  moneySchema,
  settleResponseHeader,
  SupportedEVMNetworks,
  SupportedSVMNetworks
} from "x402/types";
import { useFacilitator } from "x402/verify";
function paymentMiddleware(payTo, routes, facilitator, paywall) {
  const { verify, settle, supported } = useFacilitator(facilitator);
  const x402Version = 1;
  const routePatterns = computeRoutePatterns(routes);
  return async function paymentMiddleware2(c, next) {
    var _a;
    const method = c.req.method.toUpperCase();
    const matchingRoute = findMatchingRoute(routePatterns, c.req.path, method);
    if (!matchingRoute) {
      return next();
    }
    const { price, network, config = {} } = matchingRoute.config;
    const {
      description,
      mimeType,
      maxTimeoutSeconds,
      inputSchema,
      outputSchema,
      customPaywallHtml,
      resource,
      errorMessages,
      discoverable
    } = config;
    const atomicAmountForAsset = processPriceToAtomicAmount(price, network);
    if ("error" in atomicAmountForAsset) {
      throw new Error(atomicAmountForAsset.error);
    }
    const { maxAmountRequired, asset } = atomicAmountForAsset;
    const resourceUrl = resource || c.req.url;
    let paymentRequirements = [];
    if (SupportedEVMNetworks.includes(network)) {
      paymentRequirements.push({
        scheme: "exact",
        network,
        maxAmountRequired,
        resource: resourceUrl,
        description: description ?? "",
        mimeType: mimeType ?? "application/json",
        payTo: getAddress(payTo),
        maxTimeoutSeconds: maxTimeoutSeconds ?? 300,
        asset: getAddress(asset.address),
        // TODO: Rename outputSchema to requestStructure
        outputSchema: {
          input: {
            type: "http",
            method,
            discoverable: discoverable ?? true,
            ...inputSchema
          },
          output: outputSchema
        },
        extra: asset.eip712
      });
    } else if (SupportedSVMNetworks.includes(network)) {
      const paymentKinds = await supported();
      let feePayer;
      for (const kind of paymentKinds.kinds) {
        if (kind.network === network && kind.scheme === "exact") {
          feePayer = (_a = kind == null ? void 0 : kind.extra) == null ? void 0 : _a.feePayer;
          break;
        }
      }
      if (!feePayer) {
        throw new Error(`The facilitator did not provide a fee payer for network: ${network}.`);
      }
      paymentRequirements.push({
        scheme: "exact",
        network,
        maxAmountRequired,
        resource: resourceUrl,
        description: description ?? "",
        mimeType: mimeType ?? "",
        payTo,
        maxTimeoutSeconds: maxTimeoutSeconds ?? 60,
        asset: asset.address,
        // TODO: Rename outputSchema to requestStructure
        outputSchema: {
          input: {
            type: "http",
            method,
            ...inputSchema
          },
          output: outputSchema
        },
        extra: {
          feePayer
        }
      });
    } else {
      throw new Error(`Unsupported network: ${network}`);
    }
    const payment = c.req.header("X-PAYMENT");
    const userAgent = c.req.header("User-Agent") || "";
    const acceptHeader = c.req.header("Accept") || "";
    const isWebBrowser = acceptHeader.includes("text/html") && userAgent.includes("Mozilla");
    if (!payment) {
      if (isWebBrowser) {
        let displayAmount;
        if (typeof price === "string" || typeof price === "number") {
          const parsed = moneySchema.safeParse(price);
          if (parsed.success) {
            displayAmount = parsed.data;
          } else {
            displayAmount = Number.NaN;
          }
        } else {
          displayAmount = Number(price.amount) / 10 ** price.asset.decimals;
        }
        const currentUrl = new URL(c.req.url).pathname + new URL(c.req.url).search;
        const html = customPaywallHtml ?? getPaywallHtml({
          amount: displayAmount,
          paymentRequirements: toJsonSafe(paymentRequirements),
          currentUrl,
          testnet: network === "base-sepolia",
          cdpClientKey: paywall == null ? void 0 : paywall.cdpClientKey,
          appName: paywall == null ? void 0 : paywall.appName,
          appLogo: paywall == null ? void 0 : paywall.appLogo,
          sessionTokenEndpoint: paywall == null ? void 0 : paywall.sessionTokenEndpoint
        });
        return c.html(html, 402);
      }
      return c.json(
        {
          error: (errorMessages == null ? void 0 : errorMessages.paymentRequired) || "X-PAYMENT header is required",
          accepts: paymentRequirements,
          x402Version
        },
        402
      );
    }
    let decodedPayment;
    try {
      decodedPayment = exact.evm.decodePayment(payment);
      decodedPayment.x402Version = x402Version;
    } catch (error) {
      return c.json(
        {
          error: (errorMessages == null ? void 0 : errorMessages.invalidPayment) || (error instanceof Error ? error : new Error("Invalid or malformed payment header")),
          accepts: paymentRequirements,
          x402Version
        },
        402
      );
    }
    const selectedPaymentRequirements = findMatchingPaymentRequirements(
      paymentRequirements,
      decodedPayment
    );
    if (!selectedPaymentRequirements) {
      return c.json(
        {
          error: (errorMessages == null ? void 0 : errorMessages.noMatchingRequirements) || "Unable to find matching payment requirements",
          accepts: toJsonSafe(paymentRequirements),
          x402Version
        },
        402
      );
    }
    const verification = await verify(decodedPayment, selectedPaymentRequirements);
    if (!verification.isValid) {
      return c.json(
        {
          error: (errorMessages == null ? void 0 : errorMessages.verificationFailed) || verification.invalidReason,
          accepts: paymentRequirements,
          payer: verification.payer,
          x402Version
        },
        402
      );
    }
    await next();
    let res = c.res;
    if (res.status >= 400) {
      return;
    }
    c.res = void 0;
    try {
      const settlement = await settle(decodedPayment, selectedPaymentRequirements);
      if (settlement.success) {
        const responseHeader = settleResponseHeader(settlement);
        res.headers.set("X-PAYMENT-RESPONSE", responseHeader);
      } else {
        throw new Error(settlement.errorReason);
      }
    } catch (error) {
      res = c.json(
        {
          error: (errorMessages == null ? void 0 : errorMessages.settlementFailed) || (error instanceof Error ? error : new Error("Failed to settle payment")),
          accepts: paymentRequirements,
          x402Version
        },
        402
      );
    }
    c.res = res;
  };
}
export {
  paymentMiddleware
};
